src/rules/selector-type-case/index.js

```diff
@@ -0,0 +1,66 @@
+import selectorParser from "postcss-selector-parser"
+import {
+  cssRuleHasSelectorEndingWithColon,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-type-case"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (actual, expected) => `Expected "${actual}" to be "${expected}"`,
+})
+
+export default function (expectation) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: expectation,
+      possible: [
+        "lower",
+        "upper",
+      ],
+    })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      // Ignore keyframe selectors
+      if (rule.parent.type === "atrule" && rule.parent.name === "keyframes") {
+        return
+      }
+
+      if (cssRuleHasSelectorEndingWithColon(rule)) { return }
+
+      selectorParser(selectorAST => {
+        selectorAST.eachTag(tag => {
+          // Destructring the tag object
+          const { parent, sourceIndex, value } = tag
+             
+          // postcss-selector-parser includes the arguments to nth-child() functions
+          // as "tags", so we need to ignore them ourselves.
+          // The fake-tag's "parent" is actually a selector node, whose parent
+          // should be the :nth-child pseudo node.
+          if (parent.parent.type === "pseudo" && parent.parent.value === ":nth-child") {
+            return
+          }
+
+          // & is not a type selector: it's used for nesting 
+          if (value[0] === "&") { return }
+
+          const expectedValue = expectation === "lower" ? value.toLowerCase() : value.toUpperCase()
+
+          if (value === expectedValue) { return }
+
+          report({
+            message: messages.expected(value, expectedValue),
+            node: rule,
+            index: sourceIndex,
+            ruleName,
+            result,
+          })
+        })
+      })
+        .process(rule.selector)
```

Comment:
It's confusing to run across this after such a long function --- I couldn't remember what it was about. Can you maybe use a named instead of anonymous function and then put the selectorParser stuff together, like this?

```js
selectorParser(checkSelector).process(rule.selector)

function checkSelector(selectorAST) {
  selectorAST.eachTag(tag => {
  ..
}
```

Does that make sense?